compiler/rustc_resolve/src/imports.rs

```diff
@@ -542,31 +564,140 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
 
     // Import resolution
     //
-    // This is a fixed-point algorithm. We resolve imports until our efforts
-    // are stymied by an unresolved import; then we bail out of the current
-    // module and continue. We terminate successfully once no more imports
-    // remain or unsuccessfully when no forward progress in resolving imports
-    // is made.
+    // This is a batched fixed-point algorithm. Each import is resolved in
+    // isolation, with any side effects collected for later.
+    // After a full pass over the current set of `indeterminate_imports`,
+    // the collected side effects are committed together. The process
+    // repeats until either no imports remain or no further progress can
+    // be made.
 
     /// Resolves all imports for the crate. This method performs the fixed-
     /// point iteration.
     pub(crate) fn resolve_imports(&mut self) {
-        self.assert_speculative = true;
         let mut prev_indeterminate_count = usize::MAX;
         let mut indeterminate_count = self.indeterminate_imports.len() * 3;
         while indeterminate_count < prev_indeterminate_count {
             prev_indeterminate_count = indeterminate_count;
             indeterminate_count = 0;
+            let mut side_effects = Vec::new();
+            self.assert_speculative = true;
             for import in mem::take(&mut self.indeterminate_imports) {
-                let import_indeterminate_count = self.cm().resolve_import(import);
+                let (side_effect, import_indeterminate_count) = self.cm().resolve_import(import);
                 indeterminate_count += import_indeterminate_count;
                 match import_indeterminate_count {
                     0 => self.determined_imports.push(import),
                     _ => self.indeterminate_imports.push(import),
                 }
+                if let Some(side_effect) = side_effect {
+                    side_effects.push((import, side_effect));
+                }
+            }
+            self.assert_speculative = false;
+            self.commit_import_resolutions(side_effects);
+        }
+    }
+
+    fn commit_import_resolutions(
+        &mut self,
+        import_resolutions: Vec<(Import<'ra>, SideEffect<'ra>)>,
+    ) {
+        self.determined_imports.reserve(self.determined_imports.len());
+        for (import, side_effect) in import_resolutions.iter() {
+            self.determined_imports.push(*import);
```

Comment:
We push `determined_imports` both here and above in `resolve_imports`, that's where the duplication is coming from, most likely.